File: go-sdk/runner/exit/exit_runner.go
package exit

import (
	"github.com/konstellation-io/kre-runners/go-sdk/v1/runner/common"
	"github.com/konstellation-io/kre-runners/go-sdk/v1/sdk"

	"github.com/go-logr/logr"
	"github.com/nats-io/nats.go"
)

type Preprocessor common.Handler

type Handler common.Handler

type Postprocessor common.Handler

type Runner struct {
	sdk              sdk.KaiSDK
	nats             *nats.Conn
	jetstream        nats.JetStreamContext
	responseHandlers map[string]Handler
	initializer      common.Initializer
	preprocessor     Preprocessor
	postprocessor    Postprocessor
	finalizer        common.Finalizer
}

func NewExitRunner(logger logr.Logger, ns *nats.Conn, js nats.JetStreamContext) *Runner {
	return &Runner{
		sdk:              sdk.NewKaiSDK(logger.WithName("[Exit]"), ns, js),
		nats:             ns,
		jetstream:        js,
		responseHandlers: make(map[string]Handler),
	}
}

func (er *Runner) WithInitializer(initializer common.Initializer) *Runner {
	er.initializer = composeInitializer(initializer)
	return er
}

func (er *Runner) WithPreprocessor(preprocessor Preprocessor) *Runner {
	er.preprocessor = composePreprocessor(preprocessor)
	return er
}

func (er *Runner) WithHandler(handler Handler) *Runner {
	er.responseHandlers["default"] = composeHandler(handler)
	return er
}

func (er *Runner) WithCustomHandler(subject string, handler Handler) *Runner {
	er.responseHandlers[subject] = composeHandler(handler)
	return er
}

func (er *Runner) WithPostprocessor(postprocessor Postprocessor) *Runner {
	er.postprocessor = composePostprocessor(postprocessor)
	return er
}

func (er *Runner) WithFinalizer(finalizer common.Finalizer) *Runner {
	er.finalizer = composeFinalizer(finalizer)
	return er
}

func (er *Runner) Run() {
	if er.responseHandlers["default"] == nil {
		panic("No default handler defined")
	}
	if er.initializer == nil {
		er.initializer = composeInitializer(nil)
	}
	if er.finalizer == nil {
		er.finalizer = composeFinalizer(nil)
	}

	er.initializer(er.sdk)

	er.startSubscriber()

	er.finalizer(er.sdk)
}

File: py-sdk/runner/runner/exit/exit_runner.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

import loguru
from loguru import logger
from nats.aio.client import Client as NatsClient
from nats.js.client import JetStreamContext
from vyper import v

from runner.common.common import Finalizer, Handler, Initializer
from runner.exit.exceptions import UndefinedDefaultHandlerFunctionError
from runner.exit.helpers import (
    compose_finalizer,
    compose_handler,
    compose_initializer,
    compose_postprocessor,
    compose_preprocessor,
)
from runner.exit.subscriber import ExitSubscriber
from sdk.kai_sdk import KaiSDK

Preprocessor = Handler
Postprocessor = Handler


@dataclass
class ExitRunner:
    sdk: KaiSDK = field(init=False)
    nc: NatsClient
    js: JetStreamContext
    logger: loguru.Logger = logger.bind(context="[EXIT]")
    response_handlers: dict[str, Handler] = field(default_factory=dict)
    initializer: Optional[Initializer] = None
    preprocessor: Preprocessor = field(init=False)
    postprocessor: Postprocessor = field(init=False)
    finalizer: Optional[Finalizer] = None

    def __post_init__(self) -> None:
        self.sdk = KaiSDK(nc=self.nc, js=self.js, logger=self.logger)
        self.subscriber = ExitSubscriber(self)

    def with_initializer(self, initializer: Initializer) -> ExitRunner:
        self.initializer = compose_initializer(initializer)
        return self

    def with_preprocessor(self, preprocessor: Preprocessor) -> ExitRunner:
        self.preprocessor = compose_preprocessor(preprocessor)
        return self

    def with_handler(self, handler: Handler) -> ExitRunner:
        self.response_handlers["default"] = compose_handler(handler)
        return self

    def with_custom_handler(self, subject: str, handler: Handler) -> ExitRunner:
        self.response_handlers[subject] = compose_handler(handler)
        return self

    def with_postprocessor(self, postprocessor: Postprocessor) -> ExitRunner:
        self.postprocessor = compose_postprocessor(postprocessor)
        return self

    def with_finalizer(self, finalizer: Finalizer) -> ExitRunner:
        self.finalizer = compose_finalizer(finalizer)
        return self

    async def run(self) -> None:
        if "default" not in self.response_handlers:
            raise UndefinedDefaultHandlerFunctionError()

        if not self.initializer:
            self.initializer = compose_initializer()

        if not self.finalizer:
            self.finalizer = compose_finalizer()

        initializer_func = self.initializer(self.sdk)
        await initializer_func

        await self.subscriber.start()

        self.finalizer(self.sdk)
